[WGSL] Unify AST node type traits

djg opened the pull request:
#### 480554208313ebcb878ee7ef315bc94d9b628e0b
<pre>
[WGSL] Unify AST node type traits
<a href="https://bugs.webkit.org/show_bug.cgi?id=247475">https://bugs.webkit.org/show_bug.cgi?id=247475</a>
rdar://101943721

Reviewed by Myles C. Maxfield.

Unify WGSL AST node type introspection to use TypeCastTraits, replacing the
&quot;.isNodeKind()&quot; call pattern with &quot;is&lt;NodeKind&gt;(...)&quot; to complement
&quot;downcast&lt;NodeKind&gt;(...)&quot;.

* Source/WebGPU/WGSL/AST/AST.h:
* Source/WebGPU/WGSL/AST/ASTArrayAccess.h:
* Source/WebGPU/WGSL/AST/ASTAssignmentStatement.h:
* Source/WebGPU/WGSL/AST/ASTAttribute.h:
(WGSL::AST::Attribute::Attribute):
(isType):
(WGSL::AST::Attribute::~Attribute): Deleted.
(WGSL::AST::Attribute::isGroup const): Deleted.
(WGSL::AST::Attribute::isBinding const): Deleted.
(WGSL::AST::Attribute::isStage const): Deleted.
(WGSL::AST::Attribute::isLocation const): Deleted.
(WGSL::AST::Attribute::isBuiltin const): Deleted.
(): Deleted.
* Source/WebGPU/WGSL/AST/ASTBindingAttribute.h: Copied from Source/WebGPU/WGSL/AST/ASTGlobalDirective.h.
* Source/WebGPU/WGSL/AST/ASTBuiltinAttribute.h: Copied from Source/WebGPU/WGSL/AST/ASTGlobalDirective.h.
* Source/WebGPU/WGSL/AST/ASTCallableExpression.h:
* Source/WebGPU/WGSL/AST/ASTCompoundStatement.h:
* Source/WebGPU/WGSL/AST/ASTDecl.h:
(WGSL::AST::Decl::Decl):
(isType):
(WGSL::AST::Decl::~Decl): Deleted.
(WGSL::AST::Decl::isVariable const): Deleted.
(WGSL::AST::Decl::isStruct const): Deleted.
(WGSL::AST::Decl::isFunction const): Deleted.
* Source/WebGPU/WGSL/AST/ASTExpression.h:
(WGSL::AST::Expression::Expression):
(isTypeOf):
(WGSL::AST::Expression::~Expression): Deleted.
(WGSL::AST::Expression::isBoolLiteral const): Deleted.
(WGSL::AST::Expression::isInt32Literal const): Deleted.
(WGSL::AST::Expression::isUInt32Literal const): Deleted.
(WGSL::AST::Expression::isFloat32Literal const): Deleted.
(WGSL::AST::Expression::isAbstractIntLiteral const): Deleted.
(WGSL::AST::Expression::isAbstractFloatLiteral const): Deleted.
(WGSL::AST::Expression::isIdentifier const): Deleted.
(WGSL::AST::Expression::isArrayAccess const): Deleted.
(WGSL::AST::Expression::isStructureAccess const): Deleted.
(WGSL::AST::Expression::isCallableExpression const): Deleted.
(WGSL::AST::Expression::isUnaryExpression const): Deleted.
* Source/WebGPU/WGSL/AST/ASTFunctionDecl.h:
* Source/WebGPU/WGSL/AST/ASTGlobalDirective.h:
(WGSL::AST::GlobalDirective::GlobalDirective):
* Source/WebGPU/WGSL/AST/ASTGroupAttribute.h: Copied from Source/WebGPU/WGSL/AST/ASTGlobalDirective.h.
* Source/WebGPU/WGSL/AST/ASTIdentifierExpression.h:
* Source/WebGPU/WGSL/AST/ASTLiteralExpressions.h:
* Source/WebGPU/WGSL/AST/ASTLocationAttribute.h: Copied from Source/WebGPU/WGSL/AST/ASTGlobalDirective.h.
* Source/WebGPU/WGSL/AST/ASTNode.h:
(WGSL::AST::Node::Node):
(WGSL::AST::ASTNode::ASTNode): Deleted.
(WGSL::AST::ASTNode::span const): Deleted.
* Source/WebGPU/WGSL/AST/ASTReturnStatement.h:
* Source/WebGPU/WGSL/AST/ASTShaderModule.h:
* Source/WebGPU/WGSL/AST/ASTStageAttribute.h: Copied from Source/WebGPU/WGSL/AST/ASTGlobalDirective.h.
* Source/WebGPU/WGSL/AST/ASTStatement.h:
(WGSL::AST::Statement::Statement):
(isType):
(WGSL::AST::Statement::~Statement): Deleted.
(WGSL::AST::Statement::isCompound const): Deleted.
(WGSL::AST::Statement::isReturn const): Deleted.
(WGSL::AST::Statement::isAssignment const): Deleted.
(WGSL::AST::Statement::isVariable const): Deleted.
* Source/WebGPU/WGSL/AST/ASTStructureAccess.h:
* Source/WebGPU/WGSL/AST/ASTStructureDecl.h:
* Source/WebGPU/WGSL/AST/ASTTypeDecl.h:
(WGSL::AST::TypeDecl::TypeDecl):
(isType):
(WGSL::AST::TypeDecl::~TypeDecl): Deleted.
(WGSL::AST::TypeDecl::isArray const): Deleted.
(WGSL::AST::TypeDecl::isNamed const): Deleted.
(WGSL::AST::TypeDecl::isParameterized const): Deleted.
* Source/WebGPU/WGSL/AST/ASTUnaryExpression.h:
* Source/WebGPU/WGSL/AST/ASTVariableDecl.h:
* Source/WebGPU/WGSL/AST/ASTVariableQualifier.h:
* Source/WebGPU/WGSL/AST/ASTVariableStatement.h:
* Source/WebGPU/WGSL/AST/ASTVisitor.cpp:
(WGSL::AST::Visitor::visit):
* Source/WebGPU/WGSL/Parser.cpp:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WGSL/ConstLiteralTests.cpp:
(TestWGSLAPI::TEST):
* Tools/TestWebKitAPI/Tests/WGSL/ParserTests.cpp:
(checkBuiltin):
(checkIntLiteral):
(checkVecType):
(TestWGSLAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/256381@main">https://commits.webkit.org/256381@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6fbed137bdd02923ba8438a83896d9db546f7c3a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95394 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4668 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/28434 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/104976 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4680 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33397 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87760 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100854 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101058 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/3416 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/81993 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30486 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/85313 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87219 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73325 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39194 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/18773 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/36897 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20068 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40870 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/42725 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2119 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42857 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39309 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

litherum reviewed (approved):
Maybe this is a silly question, but should we enable RTTI instead of using all these virtual calls?

djg committed:
[WGSL] Unify AST node type traits
https://bugs.webkit.org/show_bug.cgi?id=247475
rdar://101943721

Reviewed by Myles C. Maxfield.

Unify WGSL AST node type introspection to use TypeCastTraits, replacing the
".isNodeKind()" call pattern with "is<NodeKind>(...)" to complement
"downcast<NodeKind>(...)".

* Source/WebGPU/WGSL/AST/AST.h:
* Source/WebGPU/WGSL/AST/ASTArrayAccess.h:
* Source/WebGPU/WGSL/AST/ASTAssignmentStatement.h:
* Source/WebGPU/WGSL/AST/ASTAttribute.h:
(WGSL::AST::Attribute::Attribute):
(isType):
(WGSL::AST::Attribute::~Attribute): Deleted.
(WGSL::AST::Attribute::isGroup const): Deleted.
(WGSL::AST::Attribute::isBinding const): Deleted.
(WGSL::AST::Attribute::isStage const): Deleted.
(WGSL::AST::Attribute::isLocation const): Deleted.
(WGSL::AST::Attribute::isBuiltin const): Deleted.
(): Deleted.
* Source/WebGPU/WGSL/AST/ASTBindingAttribute.h: Copied from Source/WebGPU/WGSL/AST/ASTGlobalDirective.h.
* Source/WebGPU/WGSL/AST/ASTBuiltinAttribute.h: Copied from Source/WebGPU/WGSL/AST/ASTGlobalDirective.h.
* Source/WebGPU/WGSL/AST/ASTCallableExpression.h:
* Source/WebGPU/WGSL/AST/ASTCompoundStatement.h:
* Source/WebGPU/WGSL/AST/ASTDecl.h:
(WGSL::AST::Decl::Decl):
(isType):
(WGSL::AST::Decl::~Decl): Deleted.
(WGSL::AST::Decl::isVariable const): Deleted.
(WGSL::AST::Decl::isStruct const): Deleted.
(WGSL::AST::Decl::isFunction const): Deleted.
* Source/WebGPU/WGSL/AST/ASTExpression.h:
(WGSL::AST::Expression::Expression):
(isTypeOf):
(WGSL::AST::Expression::~Expression): Deleted.
(WGSL::AST::Expression::isBoolLiteral const): Deleted.
(WGSL::AST::Expression::isInt32Literal const): Deleted.
(WGSL::AST::Expression::isUInt32Literal const): Deleted.
(WGSL::AST::Expression::isFloat32Literal const): Deleted.
(WGSL::AST::Expression::isAbstractIntLiteral const): Deleted.
(WGSL::AST::Expression::isAbstractFloatLiteral const): Deleted.
(WGSL::AST::Expression::isIdentifier const): Deleted.
(WGSL::AST::Expression::isArrayAccess const): Deleted.
(WGSL::AST::Expression::isStructureAccess const): Deleted.
(WGSL::AST::Expression::isCallableExpression const): Deleted.
(WGSL::AST::Expression::isUnaryExpression const): Deleted.
* Source/WebGPU/WGSL/AST/ASTFunctionDecl.h:
* Source/WebGPU/WGSL/AST/ASTGlobalDirective.h:
(WGSL::AST::GlobalDirective::GlobalDirective):
* Source/WebGPU/WGSL/AST/ASTGroupAttribute.h: Copied from Source/WebGPU/WGSL/AST/ASTGlobalDirective.h.
* Source/WebGPU/WGSL/AST/ASTIdentifierExpression.h:
* Source/WebGPU/WGSL/AST/ASTLiteralExpressions.h:
* Source/WebGPU/WGSL/AST/ASTLocationAttribute.h: Copied from Source/WebGPU/WGSL/AST/ASTGlobalDirective.h.
* Source/WebGPU/WGSL/AST/ASTNode.h:
(WGSL::AST::Node::Node):
(WGSL::AST::ASTNode::ASTNode): Deleted.
(WGSL::AST::ASTNode::span const): Deleted.
* Source/WebGPU/WGSL/AST/ASTReturnStatement.h:
* Source/WebGPU/WGSL/AST/ASTShaderModule.h:
* Source/WebGPU/WGSL/AST/ASTStageAttribute.h: Copied from Source/WebGPU/WGSL/AST/ASTGlobalDirective.h.
* Source/WebGPU/WGSL/AST/ASTStatement.h:
(WGSL::AST::Statement::Statement):
(isType):
(WGSL::AST::Statement::~Statement): Deleted.
(WGSL::AST::Statement::isCompound const): Deleted.
(WGSL::AST::Statement::isReturn const): Deleted.
(WGSL::AST::Statement::isAssignment const): Deleted.
(WGSL::AST::Statement::isVariable const): Deleted.
* Source/WebGPU/WGSL/AST/ASTStructureAccess.h:
* Source/WebGPU/WGSL/AST/ASTStructureDecl.h:
* Source/WebGPU/WGSL/AST/ASTTypeDecl.h:
(WGSL::AST::TypeDecl::TypeDecl):
(isType):
(WGSL::AST::TypeDecl::~TypeDecl): Deleted.
(WGSL::AST::TypeDecl::isArray const): Deleted.
(WGSL::AST::TypeDecl::isNamed const): Deleted.
(WGSL::AST::TypeDecl::isParameterized const): Deleted.
* Source/WebGPU/WGSL/AST/ASTUnaryExpression.h:
* Source/WebGPU/WGSL/AST/ASTVariableDecl.h:
* Source/WebGPU/WGSL/AST/ASTVariableQualifier.h:
* Source/WebGPU/WGSL/AST/ASTVariableStatement.h:
* Source/WebGPU/WGSL/AST/ASTVisitor.cpp:
(WGSL::AST::Visitor::visit):
* Source/WebGPU/WGSL/Parser.cpp:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WGSL/ConstLiteralTests.cpp:
(TestWGSLAPI::TEST):
* Tools/TestWebKitAPI/Tests/WGSL/ParserTests.cpp:
(checkBuiltin):
(checkIntLiteral):
(checkVecType):
(TestWGSLAPI::TEST):

Canonical link: https://commits.webkit.org/256381@main
